### PR TITLE
Fix for Issue-21 Recognize self close tags.

### PR DESCRIPTION
**Pull request by ramakrishnan**
When the HTML document is loaded in non XML mode. Cheerio should recognize all self closing tags and treat them accordingly.

### CODE REVIEW

**Comment by pradeepshastry** [comment]
:+1: Nice catch. Would definitely help

**Comment by ramakrishnan** [comment]
@fb55  may I have your comments on this PR and merge it if this makes sense.

**Comment by fb55** [comment]
In HTML 5, self closing tags are ignored. Thanks for the effort, but I don't really see a reason why this is needed.

**Comment by ramakrishnan** [comment]
Yes you are correct on HTML 5 self closing tags, But what if the document is a xhtml
When I parse it with xmlMode true it will consider all nodes which has no children into a self-closing tags. (exanple)
`<script src="something.js"></script>` to `<script src="something.js"/>`
I would like the `<script>` tag have the `</script>` and treat all self-closing tags as in xml model, a mix of both html and xml

**Comment by sandev** [comment]
Thank you Ram. I am looking for a similar fix too. Please let me know when this change is merged or we might need to fork it and use.

**Comment by pradeepshastry** [comment]
+1. I agree with Ram's point. It doesn't work well with xhtml's. Much needed fix!!

**Comment by fb55** [comment]
One idea would be to mirror htmlparser2's parsing behavior and to have this functionality available via the `recognizeSelfClosing` option. `<script />` is still valid in XHTML though, I still don't see the use case.

**Comment by ramakrishnan** [comment]
How about having an extra option to cheerio  htmlWithSelfClosing for this fix. This will be useful for files which have self-closing tags in them and cheerio can retain them as is.

**Comment by ramakrishnan** [comment]
@fb55 
Any thoughts on my earlier suggestions. It would be a nice to have feature if cheerio could keep the html  xhtml tags initial formatting as is.

**Comment by fb55** [comment]
How do you intend to detect XHTML? I assume that recognizeSelfClosing is
part of it, mirroring it in the serializer should work.

**Comment by ramakrishnan** [comment]
@fb55 
Do we mean to start using recognizeSelfClosing in dom-serializer which will ensure all self closing tags will have a trailing `/` to them?

**Comment by wilsonwc** [comment]
We have a use case for this as well, but my thought is that it may be better to turn it on/off based on a flag like xhtmlMode since right now the self closing tags get applied to all html doctypes.  I think it would be reasonable for the user to have to set this option properly when they load the file into cheerio.

**Comment by ramakrishnan** [comment]
@wilsonwc Thanks a lot for your comment, even I am thinking on the same line. It will be really nice to have a flag which can add self closing tags. But I am waiting for conformation from @fb55 to decide on the flag usage and to make further changes.

**Comment by pradeepshastry** [comment]
@fb55 : Can we get an update on this. This would help me solve a lot of issues in my current project. We have a similar situation like this

File: index.js [hunk 1]
```diff
@@ -141,7 +141,11 @@ function renderTag(elem, opts) {
   ) {
     tag += '/>';
   } else {
-    tag += '>';
+    if (singleTag[elem.name] && opts.recognizeSelfClosing) {
```
Comment (fb55) [inline index.js]: Should be integrated in the previous `if` statement.

**Comment by fb55** [comment]
I'm generally okay with the change, just one nitpick (see comment) and this can be merged.

File: index.js (hunk 1 above)
Comment (ramakrishnan) [inline index.js]: `opts.recognizeSelfClosing` and `opts.xmlModel` cannot be together, hence the first `if` will lead to the `else` block.

Am I missing anything?

Comment (fb55) [inline index.js]: No, I simply dislike having two identical branches. I would also opt for reusing the children check from the previous branch, as that also includes tags wich simply lack any content.

(The change would then consist of replacing `opts.xmlMode` with `(opts.xmlMode || opts.recognizeSelfClosing)`.)

File: index.js [hunk 2]
```diff
@@ -136,8 +136,9 @@ function renderTag(elem, opts) {
   }
 
   if (
-    opts.xmlMode
-    && (!elem.children || elem.children.length === 0)
+    (opts.xmlMode
+    && (!elem.children || elem.children.length === 0)) ||
```
Comment (fb55) [inline index.js]: Any reason not to reuse this check?

Comment (ramakrishnan) [inline index.js]: These are to be separate checks set `tag += '/>'` when
`opts.xmlMode   && (!elem.children || elem.children.length === 0)` matches or when `singleTag[elem.name] && opts.recognizeSelfClosing`

Comment (fb55) [inline index.js]: In other words: Why do you check for `singleTag[elem.name]`, and not for `!elem.children || elem.children.length === 0`?

Comment (ramakrishnan) [inline index.js]: Ideally only self closing tags which are listed under `singleTag` should have self closing. Moving the check along with `!elem.children || elem.children.length === 0` will make any element which has no children to be self closing. Example
`<iframe src="test"></iframe>` will become `<iframe src="test"/>`

Comment (ramakrishnan) [inline index.js]: @fb55 Is everything OK with this PR?

Comment (fb55) [inline index.js]: I would prefer consistent behavior, as for `xmlMode`. When you're allowing self-closing tags in the first place, all tags should be treated this way. (There is no HTML/XML/XHTML version that allows self-closing tags solely for void elements.)

Comment (ramakrishnan) [inline index.js]: This change has no connection to `xmlMode` that is why I have a `||` condition. Tags which list under `singleTag` will have a `/>`

Comment (ramakrishnan) [inline index.js]: @fb55 Do you have concerns in this pull request.

Comment (pradeepshastry) [inline index.js]: It will be very helpful to have this feature. It will be a icing on the cake for a module which makes any developers life easier.

Comment (ramakrishnan) [inline index.js]: @fb55 Do you have concerns in this pull request.

Comment (ramakrishnan) [inline index.js]: @fb55 Do you have concerns in this pull request.

Comment (ramakrishnan) [inline index.js]: @fb55 A gentle reminder.

**Comment by ramakrishnan** [comment]
@fb55 Can we get this pull request merged.

**Comment by pradeepshastry** [comment]
@fb55 - It would really be of great help to have this feature merged into such an awesome library. It will definitely benefit a lot of folks.

**Comment by fb55** [comment]
Closing this, because

- The way this PR was initially presented feels very sketchy (another person who's associated with the author immediately weighing in, without mentioning their connection).
- My feedback concerning consistent behavior was disregarded – although I wrote *"I would prefer"*, that still was an opportunity to either convince me or apply the necessary changes.
- This only partially fixes the problem at hand.

If someone wants to open another PR that provides consistent behavior, I'm open to discussing this again.

**Pull request closed without merging**